src/main/scala/com/advancedtelematic/campaigner/db/Campaigns.scala

```diff
@@ -150,25 +156,88 @@ protected [db] class Campaigns(implicit db: Database, ec: ExecutionContext)
       }
     }
 
-  // TODO: I think this can go, see schema.scala
-  private def findGroups(ns: Namespace, campaign: CampaignId): Future[Set[GroupId]] =
-    db.run {
-      campaignRepo.findAction(ns, campaign).flatMap { _ =>
+  private def findGroupsAction(campaignId: CampaignId): DBIO[Set[GroupId]] =
+    campaignRepo.findAction(campaignId).flatMap { _ =>
         Schema.campaignGroups
-          .filter(_.campaignId === campaign)
+          .filter(_.campaignId === campaignId)
           .map(_.groupId)
           .result
           .map(_.toSet)
-      }
     }
+}
+
+protected [db] class CampaignStatusTransition(implicit db: Database, ec: ExecutionContext) extends CampaignSupport
+  with GroupStatsSupport with CancelTaskSupport  {
+
+  def devicesFinished(campaignId: CampaignId): DBIO[Unit] =
+    updateToCalculatedStatus(campaignId)
+
+  def launch(campaignId: CampaignId): DBIO[CampaignId] =
+    campaignRepo.setStatusAction(campaignId, CampaignStatus.launched)
+
+  def cancel(campaignId: CampaignId): DBIO[Unit] =
```

Comment:
Maybe `cancelAction` for consistency with all the other `cancelAction`s ?

Reply:
This all seems rather arbitrary but I don't think it is. *Action methods are only used internally in this class, because this class only exposes DBIO Actions, it never exposes `Future[_]`.  

Where I think the confusion comes from, there are *Action methods for repositories, but those classes expose both DBIO and Future, so the method names have *Action to differentiate from the methods that expose Future.

Reply:
Oh I see. Yes, that was were my confusion was coming from. Thanks for the clarification :)